### ISSUE TITLE
[JENKINS-46336] Sidecar docker containers for declarative

### ISSUE THREAD

**Issue by jenkins-infra-bot** [jenkinsci/pipeline-model-definition-plugin#1137]
<p>The "sidecar container" pattern is a popular one for setting up dependent containers for running tests/builds. </p>

<p> </p>

<p>It would be nice to have support for this pattern in declarative to avoid the need to drop to script. </p>

<p> </p>

<p>To investigate...</p>

<p> </p>

<p>In scope: </p>
<ul>
	<li>Be able to specify sidecar containers, their network/port sharing with main build, in agent section</li>
	<li>Cleanup of containers</li>
	<li>Formalise the patterns shown: <a href="https://jenkins.io/doc/book/pipeline/docker/#running-sidecar-containers" class="external-link" target="_blank" rel="nofollow noopener">https://jenkins.io/doc/book/pipeline/docker/#running-sidecar-containers</a> in declarative, essentially</li>
</ul>

<p>testing: </p>
<ul>
	<li>Should be able to be fully automated using ci.jenkins.io </li>
	<li>check sidecar containers are shut down in all scenarios possible</li>
</ul>

---
<details><summary><i>Originally reported by <img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/michaelneale.png" title="michaelneale's avatar" /> <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=michaelneale">michaelneale</a>, imported from: <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-46336" target="_blank">Sidecar docker containers for declarative</a></i></summary>
<i><ul>
<li><b>status</b>: Open
<li><b>priority</b>: Major
<li><b>component(s)</b>: pipeline-model-definition-plugin
<li><b>resolution</b>: Unresolved
<li><b>votes</b>: 32
<li><b>watchers</b>: 48
<li><b>imported</b>: 2025-12-06
</ul></i>
<details><summary>Raw content of original issue</summary>

<pre>
<p>The "sidecar container" pattern is a popular one for setting up dependent containers for running tests/builds. </p>

<p> </p>

<p>It would be nice to have support for this pattern in declarative to avoid the need to drop to script. </p>

<p> </p>

<p>To investigate...</p>

<p> </p>

<p>In scope: </p>
<ul>
	<li>Be able to specify sidecar containers, their network/port sharing with main build, in agent section</li>
	<li>Cleanup of containers</li>
	<li>Formalise the patterns shown: <a href="https://jenkins.io/doc/book/pipeline/docker/#running-sidecar-containers" class="external-link" target="_blank" rel="nofollow noopener">https://jenkins.io/doc/book/pipeline/docker/#running-sidecar-containers</a> in declarative, essentially</li>
</ul>

<p>testing: </p>
<ul>
	<li>Should be able to be fully automated using ci.jenkins.io </li>
	<li>check sidecar containers are shut down in all scenarios possible</li>
</ul>
</pre>
</details>
</details>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/michaelneale.png" title="michaelneale's avatar" /> michaelneale:</i></summary>

<ul>
<li><i><a class="no-jira-link-rewrite" href="https://issue-redirect.jenkins.io/issue/46336?focusedId=316038&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-316038">Comment link</a></i>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-46336?focusedId=316038&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-316038">comment link</a> (no redirect)</i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>just bumped priority mostly as it is not a small amount of effort, some planning needed </p>
</pre>
</details>

</ul>
</details>

<p>just bumped priority mostly as it is not a small amount of effort, some planning needed </p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/jamesdumay.png" title="jamesdumay's avatar" /> jamesdumay:</i></summary>

<ul>
<li><i><a class="no-jira-link-rewrite" href="https://issue-redirect.jenkins.io/issue/46336?focusedId=316118&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-316118">Comment link</a></i>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-46336?focusedId=316118&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-316118">comment link</a> (no redirect)</i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=michaelneale" class="user-hover" rel="michaelneale">michaelneale</a> how much of this is just replicating docker compose? Perhaps we should be able to define a docker compose yml?</p>
</pre>
</details>

</ul>
</details>

<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=michaelneale" class="user-hover" rel="michaelneale">michaelneale</a> how much of this is just replicating docker compose? Perhaps we should be able to define a docker compose yml?</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/michaelneale.png" title="michaelneale's avatar" /> michaelneale:</i></summary>

<ul>
<li><i><a class="no-jira-link-rewrite" href="https://issue-redirect.jenkins.io/issue/46336?focusedId=316119&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-316119">Comment link</a></i>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-46336?focusedId=316119&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-316119">comment link</a> (no redirect)</i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=jamesdumay" class="user-hover" rel="jamesdumay">jamesdumay</a> good question - I think docker compose may do a poor job at cleaning up after if things went wrong, btu that may be one way to implement this yes (although it does require they install it, and not sure how widely it is used). </p>
</pre>
</details>

</ul>
</details>

<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=jamesdumay" class="user-hover" rel="jamesdumay">jamesdumay</a> good question - I think docker compose may do a poor job at cleaning up after if things went wrong, btu that may be one way to implement this yes (although it does require they install it, and not sure how widely it is used). </p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/jamesdumay.png" title="jamesdumay's avatar" /> jamesdumay:</i></summary>

<ul>
<li><i><a class="no-jira-link-rewrite" href="https://issue-redirect.jenkins.io/issue/46336?focusedId=316123&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-316123">Comment link</a></i>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-46336?focusedId=316123&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-316123">comment link</a> (no redirect)</i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>ack</p>
</pre>
</details>

</ul>
</details>

<p>ack</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/michaelneale.png" title="michaelneale's avatar" /> michaelneale:</i></summary>

<ul>
<li><i><a class="no-jira-link-rewrite" href="https://issue-redirect.jenkins.io/issue/46336?focusedId=316124&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-316124">Comment link</a></i>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-46336?focusedId=316124&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-316124">comment link</a> (no redirect)</i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=jamesdumay" class="user-hover" rel="jamesdumay">jamesdumay</a> other tools tend to build in sidecar functionality, probably for a reason like this... </p>
</pre>
</details>

</ul>
</details>

<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=jamesdumay" class="user-hover" rel="jamesdumay">jamesdumay</a> other tools tend to build in sidecar functionality, probably for a reason like this... </p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/orrc.svg" title="orrc's avatar" /> orrc:</i></summary>

<ul>
<li><i><a class="no-jira-link-rewrite" href="https://issue-redirect.jenkins.io/issue/46336?focusedId=331077&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-331077">Comment link</a></i>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-46336?focusedId=331077&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-331077">comment link</a> (no redirect)</i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Duplicate of <a href="https://issue-redirect.jenkins.io/issue/41581" title="Declarative: Support for multiple docker agents run in parallel and communicating" class="issue-link" data-issue-key="JENKINS-41581"><del>JENKINS-41581</del></a>?</p>
</pre>
</details>

</ul>
</details>

<p>Duplicate of <a href="https://issue-redirect.jenkins.io/issue/41581" title="Declarative: Support for multiple docker agents run in parallel and communicating" class="issue-link" data-issue-key="JENKINS-41581"><del>JENKINS-41581</del></a>?</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/jamesdumay.png" title="jamesdumay's avatar" /> jamesdumay:</i></summary>

<ul>
<li><i><a class="no-jira-link-rewrite" href="https://issue-redirect.jenkins.io/issue/46336?focusedId=331138&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-331138">Comment link</a></i>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-46336?focusedId=331138&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-331138">comment link</a> (no redirect)</i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=orrc" class="user-hover" rel="orrc">orrc</a> good eye <img class="emoticon" src="https://issues.jenkins.io/images/icons/emoticons/smile.png" height="16" width="16" align="absmiddle" alt="" border="0"/> Thank you!</p>
</pre>
</details>

</ul>
</details>

<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=orrc" class="user-hover" rel="orrc">orrc</a> good eye <img class="emoticon" src="https://issues.jenkins.io/images/icons/emoticons/smile.png" height="16" width="16" align="absmiddle" alt="" border="0"/> Thank you!</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/csanchez.png" title="csanchez's avatar" /> csanchez:</i></summary>

<ul>
<li><i><a class="no-jira-link-rewrite" href="https://issue-redirect.jenkins.io/issue/46336?focusedId=334336&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-334336">Comment link</a></i>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-46336?focusedId=334336&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-334336">comment link</a> (no redirect)</i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>There is an implementation for Kubernetes in review at  <a href="https://github.com/jenkinsci/kubernetes-plugin/pull/306" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/jenkinsci/kubernetes-plugin/pull/306</a> (<a href="https://issue-redirect.jenkins.io/issue/48135" title="kubernetes-plugin should support multiple containers in declarative templates" class="issue-link" data-issue-key="JENKINS-48135"><del>JENKINS-48135</del></a>)</p>
</pre>
</details>

</ul>
</details>

<p>There is an implementation for Kubernetes in review at  <a href="https://github.com/jenkinsci/kubernetes-plugin/pull/306" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/jenkinsci/kubernetes-plugin/pull/306</a> (<a href="https://issue-redirect.jenkins.io/issue/48135" title="kubernetes-plugin should support multiple containers in declarative templates" class="issue-link" data-issue-key="JENKINS-48135"><del>JENKINS-48135</del></a>)</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/csanchez.png" title="csanchez's avatar" /> csanchez:</i></summary>

<ul>
<li><i><a class="no-jira-link-rewrite" href="https://issue-redirect.jenkins.io/issue/46336?focusedId=338829&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-338829">Comment link</a></i>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-46336?focusedId=338829&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-338829">comment link</a> (no redirect)</i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>kubernetes plugin now supports Pods in declarative, using yaml definition</p>
</pre>
</details>

</ul>
</details>

<p>kubernetes plugin now supports Pods in declarative, using yaml definition</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/davidvanlaatum.png" title="davidvanlaatum's avatar" /> davidvanlaatum:</i></summary>

<ul>
<li><i><a class="no-jira-link-rewrite" href="https://issue-redirect.jenkins.io/issue/46336?focusedId=343544&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-343544">Comment link</a></i>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-46336?focusedId=343544&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-343544">comment link</a> (no redirect)</i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Here is a way of doing it until proper support is added</p>
<div class="code panel" style="border-width: 1px;"><div class="codeContent panelContent">
<pre class="code-java">    stages {
        stage(<span class="code-quote">'Run Tests'</span>) {
            parallel {
                stage(<span class="code-quote">"RSpec"</span>) {
                  agent { label <span class="code-quote">'docker'</span> }
                  steps {
                    script {
                      docker.image(<span class="code-quote">'postgres'</span>).withRun(<span class="code-quote">'-e POSTGRES_DB=xxxx_test'</span>) { container -&gt;
                        docker.build(<span class="code-quote">'netadmin-jenkins'</span>,<span class="code-quote">' --build-arg http_proxy=$http_proxy --build-arg https_proxy=$https_proxy --build-arg no_proxy=$no_proxy -f Dockerfile.jenkins --pull .'</span>).inside(<span class="code-quote">"--link=${container.id}:postgres -e DATABASE_URL=postgresql:<span class="code-comment">//postgres:postgres@postgres:5432/xxxx_test"</span>) {
</span>                          withEnv([<span class="code-quote">'DATABASE_URL=postgres:<span class="code-comment">//postgres@postgres:5432/'</span>,<span class="code-quote">'PGHOST=postgres'</span>,<span class="code-quote">'PGUSER=postgres'</span>,<span class="code-quote">'RAILS_ENV=test'</span>]) {
</span>                              sh <span class="code-quote">'yarn install'</span>
                              rake <span class="code-quote">'db:create'</span>
                              rake <span class="code-quote">'db:migrate'</span>
                              rake <span class="code-quote">'db:seed'</span>
                              rake <span class="code-quote">'spec'</span>
                          }
                        }
                      }
                    }
                  }
                  post {
                      always {
                          junit <span class="code-quote">'**/reports/*.xml'</span>
                      }
                  }
                }
                stage(<span class="code-quote">"RuboCop"</span>) {
                    agent {
                        dockerfile {
                          additionalBuildArgs <span class="code-quote">'--pull --build-arg http_proxy=$http_proxy --build-arg https_proxy=$https_proxy --build-arg no_proxy=$no_proxy'</span>
                          filename <span class="code-quote">'Dockerfile.jenkins'</span>
                        }
                    }
                    steps {
                        rake <span class="code-quote">'rubocop | tee rubocop.log'</span>
                    }
                    post {
                        always {
                            warnings canResolveRelativePaths: <span class="code-keyword">false</span>, failedTotalHigh: <span class="code-quote">'0'</span>, healthy: <span class="code-quote">'0'</span>, parserConfigurations: [[parserName: <span class="code-quote">'RuboCop'</span>, pattern: <span class="code-quote">'rubocop.log'</span>]], unHealthy: <span class="code-quote">'100'</span>, unstableTotalAll: <span class="code-quote">'0'</span>
                        }
                    }
                }
</pre>
</div></div>
</pre>
</details>

</ul>
</details>

<p>Here is a way of doing it until proper support is added</p>

<pre>
    stages {
stage(<span class="code-quote">'Run Tests'</span>) {
    parallel {
stage(<span class="code-quote">"RSpec"</span>) {
  agent { label <span class="code-quote">'docker'</span> }
  steps {
    script {
      docker.image(<span class="code-quote">'postgres'</span>).withRun(<span class="code-quote">'-e POSTGRES_DB=xxxx_test'</span>) { container ->
docker.build(<span class="code-quote">'netadmin-jenkins'</span>,<span class="code-quote">' --build-arg http_proxy=$http_proxy --build-arg https_proxy=$https_proxy --build-arg no_proxy=$no_proxy -f Dockerfile.jenkins --pull .'</span>).inside(<span class="code-quote">"--link=${container.id}:postgres -e DATABASE_URL=postgresql:<span class="code-comment">//postgres:postgres@​postgres:5432/xxxx_test"</span>) {
</span>  withEnv([<span class="code-quote">'DATABASE_URL=postgres:<span class="code-comment">//postgres@​postgres:5432/'</span>,<span class="code-quote">'PGHOST=postgres'</span>,<span class="code-quote">'PGUSER=postgres'</span>,<span class="code-quote">'RAILS_ENV=test'</span>]) {
</span>      sh <span class="code-quote">'yarn install'</span>
      rake <span class="code-quote">'db:create'</span>
      rake <span class="code-quote">'db:migrate'</span>
      rake <span class="code-quote">'db:seed'</span>
      rake <span class="code-quote">'spec'</span>
  }
}
      }
    }
  }
  post {
      always {
  junit <span class="code-quote">'**/reports/*.xml'</span>
      }
  }
}
stage(<span class="code-quote">"RuboCop"</span>) {
    agent {
dockerfile {
  additionalBuildArgs <span class="code-quote">'--pull --build-arg http_proxy=$http_proxy --build-arg https_proxy=$https_proxy --build-arg no_proxy=$no_proxy'</span>
  filename <span class="code-quote">'Dockerfile.jenkins'</span>
}
    }
    steps {
rake <span class="code-quote">'rubocop | tee rubocop.log'</span>
    }
    post {
always {
    warnings canResolveRelativePaths: <span class="code-keyword">false</span>, failedTotalHigh: <span class="code-quote">'0'</span>, healthy: <span class="code-quote">'0'</span>, parserConfigurations: [[parserName: <span class="code-quote">'RuboCop'</span>, pattern: <span class="code-quote">'rubocop.log'</span>]], unHealthy: <span class="code-quote">'100'</span>, unstableTotalAll: <span class="code-quote">'0'</span>
}
    }
}
</pre>

**Comment by jenkins-infra-bot** [comment]
<i>[Original `is_duplicated_by` from Jira: <a href="https://github.com/jenkinsci/pipeline-model-definition-plugin/issues?q=is%3Aissue%20%22jira_issue_key%3DJENKINS-41581%22">JENKINS-41581</a>]</i>

**Comment by jenkins-infra-bot** [comment]
<i>[Original `relates_to` from Jira: <a href="https://github.com/jenkinsci/pipeline-model-definition-plugin/issues?q=is%3Aissue%20%22jira_issue_key%3DJENKINS-48135%22">JENKINS-48135</a>]</i>

**Comment by jenkins-infra-bot** [comment]
<i>[Original `epic` from Jira: <a href="https://github.com/jenkinsci/pipeline-model-definition-plugin/issues?q=is%3Aissue%20%22jira_issue_key%3DJENKINS-47962%22">JENKINS-47962</a>]</i>
